packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts

```diff
@@ -0,0 +1,246 @@
+import type { Knex } from 'knex';
+import createDebug from 'debug';
+import type { Migration } from '../common';
+import type { Metadata } from '../../metadata';
+import { type Database, type MetadataOptions } from '../..';
+import { getUnshortenedName } from '../../utils/identifiers/shortener';
+
+const debug = createDebug('strapi::database::migration');
+
+type NameDiff<T> = {
+  short: T;
+  full: T;
+};
+
+// TODO: key isn't really needed except for debugging, we can remove it
+type IdentifierDiffs = {
+  indexes: NameDiff<{ index: number; key: string; tableName: string; indexName: string }>[];
+  tables: NameDiff<{ index: number; key: string; tableName: string }>[];
+  columns: NameDiff<{ index: number; key: string; tableName: string; columnName: string }>[];
+};
+
+export const renameIdentifiersLongerThanMaxLength: Migration = {
+  name: '5.0.0-rename-identifiers-longer-than-max-length',
+  async up(knex, db) {
+    const md = db.metadata;
+
+    const maxLength = db.DEFAULT_MAX_IDENTIFIER_LENGTH;
+
+    const metadataOptions = {
+      maxLength,
+    };
+
+    const diffs = findDiffs(md, metadataOptions);
+    // migrate indexes before tables so we know to target the original tableName
+    for (const indexDiff of diffs.indexes) {
+      await renameIndex(knex, db, indexDiff);
+    }
+
+    // migrate columns before table names so we know to target the original tableName
+    for (const columnDiff of diffs.columns) {
+      const { full, short } = columnDiff;
+      const tableName = full.tableName;
+
+      const hasTable = await knex.schema.hasTable(tableName);
+
+      if (hasTable) {
+        await knex.schema.table(tableName, async (table) => {
+          const hasColumn = await knex.schema.hasColumn(tableName, full.columnName);
+
+          if (hasColumn) {
+            debug(`renaming column ${full.columnName} to ${short.columnName}`);
+            table.renameColumn(full.columnName, short.columnName);
+          }
+        });
+      }
+    }
+
+    // migrate table names
+    for (const tableDiff of diffs.tables) {
+      const hasTable = await knex.schema.hasTable(tableDiff.full.tableName);
+
+      if (hasTable) {
+        debug(`renaming table ${tableDiff.full.tableName} to ${tableDiff.short.tableName}`);
+        await knex.schema.renameTable(tableDiff.full.tableName, tableDiff.short.tableName);
+      }
+    }
+  },
+  async down() {
+    throw new Error('not implemented');
+  },
+};
+
+type IndexDiff = NameDiff<{ index: number; key: string; tableName: string; indexName: string }>;
```

Comment:
Why not reusing `IdentifiersDiff['indexes']`?

Reply:
It was an array, but I cleaned up that it was a duplicate